hub/graphql/mutations.py

```diff
@@ -242,89 +220,68 @@ class ExternalDataSourceInput:
+    actionnetwork: Optional[ActionNetworkSourceInput] = None
 
 
 @strawberry.type
 class CreateExternalDataSourceOutput(MutationOutput):
     result: Optional[model_types.ExternalDataSource]
 
 
-def create_airtable_source(
-    info: Info, data: CreateExternalDataSourceInput
-) -> CreateExternalDataSourceOutput:
-    return get_or_create_with_computed_args(
-        models.AirtableSource,
-        info,
-        find_filter={
-            "api_key": data.airtable.api_key,
-            "base_id": data.airtable.base_id,
-            "table_id": data.airtable.table_id,
-        },
-        data=data.airtable,
-        computed_args=lambda info, data, model: {
-            "organisation": get_or_create_organisation_for_source(info, data)
-        },
-    )
-
-
-def create_mailchimp_source(
-    info: Info, data: CreateExternalDataSourceInput
-) -> models.ExternalDataSource:
-    return get_or_create_with_computed_args(
-        models.MailchimpSource,
-        info,
-        find_filter={
-            "api_key": data.mailchimp.api_key,
-            "list_id": data.mailchimp.list_id,
-        },
-        data=data.mailchimp,
-        computed_args=lambda info, data, model: {
-            "organisation": get_or_create_organisation_for_source(info, data)
-        },
-    )
-
-
 @strawberry_django.mutation(extensions=[IsAuthenticated()])
 def create_external_data_source(
     info: Info, input: CreateExternalDataSourceInput
 ) -> models.ExternalDataSource:
-    source_creators = {
-        "airtable": create_airtable_source,
-        "mailchimp": create_mailchimp_source,
-    }
-
+    input_dict = graphql_type_to_dict(input, delete_null_keys=True)
     creator_fn = None
-    for key, fn in source_creators.items():
-        source_input = getattr(input, key, None)
-        if source_input is not None:
-            creator_fn = fn
+    for crm_type_key, model in models.source_models.items():
+        if crm_type_key in input_dict and input_dict[crm_type_key] is not None:
+            kwargs = input_dict[crm_type_key]
+            kwargs["organisation"] = kwargs.get(
+                "organisation", None
+            ) or get_or_create_organisation_for_user(info)
+            print(f"Creating source of type {crm_type_key}", kwargs)
+
+            def creator_fn():  # noqa: F811
+                return model.objects.create(**kwargs)
+
+            break
 
     if creator_fn is None:
         return CreateExternalDataSourceOutput(
             code=400,
-            errors=["You must provide input data for a specific source type"],
+            errors=[
+                MutationError(
+                    code=400,
+                    message=("You must provide input data for a specific source type."),
+                )
+            ],
             result=None,
         )
 
     try:
-        result: tuple[models.ExternalDataSource, bool] = creator_fn(info, input)
-        (source, created) = result
+        source = creator_fn()
```

Comment:
question: because this is `create` instead of `get_or_create`, we lose the functionality of saying "This data source already exists". Should we restore it?

Reply:
Think we should!